Support gPTP sync with PI control and fix i.MXRT1060 support

yangbolu1991 opened the pull request:
I'm just starting to look at zephyr gPTP with a i.MXRT1060 EVK board on hand.
It looks like there are many problems in both drivers and gPTP stack.
So, I basically fixed key issues in ENET MAC/PTP drivers for RT1060. Also added a simple PI control algorithm to support continuous frequency adjustment in gPTP stack.

Finally the sync offset convergence on RT1060 was as below log. It's stable at about 100ns sync offset. (RT1060 PTP clock nominal frequency is 25M, 40ns period)

```
[00:00:02.822,000] <inf> phy_mc_ksz8081: PHY 0 is up
[00:00:02.822,000] <inf> phy_mc_ksz8081: PHY (0) Link speed 100 Mb, full duplex

[00:00:02.823,000] <inf> eth_nxp_enet_mac: Link is up
*** Booting Zephyr OS build v4.1.0-3062-g764340614802 ***
[00:00:02.826,000] <inf> net_config: Initializing network
[00:00:02.826,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:02.928,000] <inf> net_config: IPv6 address: 2001:db8::1
[00:00:03.830,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:05.268,000] <dbg> net_gptp_sample: gptp_phase_dis_cb: GM FA:FA:62:FF:FE:FE:A2:11 last phase 0.0
[00:00:06.893,000] <inf> net_gptp: sync offset   -117781 ns, freq offset -117781.000000 ppb
[00:00:07.894,000] <inf> net_gptp: sync offset    -73643 ns, freq offset -108977.300000 ppb
[00:00:08.894,000] <inf> net_gptp: sync offset    -37936 ns, freq offset -95363.200000 ppb
[00:00:09.894,000] <inf> net_gptp: sync offset    -15866 ns, freq offset -84674.000000 ppb
[00:00:10.895,000] <inf> net_gptp: sync offset     -4318 ns, freq offset -77885.800000 ppb
[00:00:11.895,000] <inf> net_gptp: sync offset       564 ns, freq offset -74299.200000 ppb
[00:00:12.895,000] <inf> net_gptp: sync offset      1731 ns, freq offset -72963.000000 ppb
[00:00:13.895,000] <inf> net_gptp: sync offset      1687 ns, freq offset -72487.700000 ppb
[00:00:14.896,000] <inf> net_gptp: sync offset      1098 ns, freq offset -72570.600000 ppb
[00:00:15.896,000] <inf> net_gptp: sync offset       540 ns, freq offset -72799.200000 ppb
[00:00:16.896,000] <inf> net_gptp: sync offset       181 ns, freq offset -72996.200000 ppb
[00:00:17.896,000] <inf> net_gptp: sync offset       127 ns, freq offset -72995.900000 ppb
[00:00:18.896,000] <inf> net_gptp: sync offset       -50 ns, freq offset -73134.800000 ppb
[00:00:19.897,000] <inf> net_gptp: sync offset        59 ns, freq offset -73040.800000 ppb
[00:00:20.897,000] <inf> net_gptp: sync offset       -39 ns, freq offset -73121.100000 ppb
[00:00:21.897,000] <inf> net_gptp: sync offset        41 ns, freq offset -73052.800000 ppb
[00:00:22.897,000] <inf> net_gptp: sync offset       -89 ns, freq offset -73170.500000 ppb
[00:00:23.897,000] <inf> net_gptp: sync offset        62 ns, freq offset -73046.200000 ppb
[00:00:24.897,000] <inf> net_gptp: sync offset       -64 ns, freq offset -73153.600000 ppb
[00:00:25.898,000] <inf> net_gptp: sync offset       -27 ns, freq offset -73135.800000 ppb
[00:00:26.898,000] <inf> net_gptp: sync offset        50 ns, freq offset -73066.900000 ppb
[00:00:27.898,000] <inf> net_gptp: sync offset        -3 ns, freq offset -73104.900000 ppb
[00:00:28.899,000] <inf> net_gptp: sync offset        66 ns, freq offset -73036.800000 ppb
[00:00:29.899,000] <inf> net_gptp: sync offset       -63 ns, freq offset -73146.000000 ppb
[00:00:30.899,000] <inf> net_gptp: sync offset       -28 ns, freq offset -73129.900000 ppb
[00:00:31.900,000] <inf> net_gptp: sync offset       132 ns, freq offset -72978.300000 ppb
[00:00:32.900,000] <inf> net_gptp: sync offset        -7 ns, freq offset -73077.700000 ppb
[00:00:33.900,000] <inf> net_gptp: sync offset      -139 ns, freq offset -73211.800000 ppb
[00:00:34.901,000] <inf> net_gptp: sync offset       -58 ns, freq offset -73172.500000 ppb
[00:00:35.901,000] <inf> net_gptp: sync offset         1 ns, freq offset -73130.900000 ppb
[00:00:36.901,000] <inf> net_gptp: sync offset        82 ns, freq offset -73049.600000 ppb
[00:00:37.901,000] <inf> net_gptp: sync offset        -4 ns, freq offset -73111.000000 ppb
[00:00:38.902,000] <inf> net_gptp: sync offset       109 ns, freq offset -72999.200000 ppb

```

Thanks.

yangbolu1991 committed:
drivers: eth_nxp_enet: fix data share with ptp driver

The enet handle in mac driver was not shared with ptp driver
properly. This was causing wrong TX timestamp.
This patch is to fix it.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ptp_clock_nxp_enet: adjust rate based on nominal frequency

The rate adjustment should be based on nomianl frequency, but not
current frequency. Then any PTP stack with PID control could adjust
frequency well.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: clock_control_mcux_ccm: use fixed 25M for PTP on RT10XX

The RT10XX uses fixed 25M for PTP clock per RM. Verified on
RT1060.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: gptp: adjust only frequency for continuous synchronization

Current clock synchronization was always stepping clock. This was
causing large offset, and discontiguous ptp hardware clock time.
For TSN hardware, discontiguous ptp hardware clock time was not
able to be used for other TSN protocols.

This patch is to convert to frequency adjustment with a basic
PI control algorithm.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ptp_clock_nxp_enet: avoid configuring IRQ handlers again

Converted ENET_Ptp1588Configure to ENET_Ptp1588StartTimer during reset.
This is to avoid configuring IRQ handlers again in hal driver with
ENET_Ptp1588Configure.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_mi.c:
```diff
@@ -837,13 +835,17 @@ static void gptp_update_local_port_clock(void)
+		 * the real time convergence of synchronization offset.
+		 * This is what users care about.
+		 */
+		if (0) {
```
I suggest creating a Kconfig symbol for this so that user does not need to adjust the source code to see the prints.
This could be named for example `CONFIG_NET_GPTP_MONITOR_SYNC_DIFF` or similar.
Then you could do here
```
if (IS_ENABLED(CONFIG_NET_GPTP_MONITOR_SYNC_DIFF)) {
....
}
```

yangbolu1991 replied:
Thanks a lot, will do that.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_mi.c:
```diff
@@ -837,13 +835,17 @@ static void gptp_update_local_port_clock(void)
 
-		ptp_clock_adjust(clk, nanosecond_diff);
+		ptp_clock_rate_adjust(clk, 1.0 + (ppb / 1000000000.0));
+		/* This prints continuous messages but can be enabled to see
```
Please add empty line before this comment

yangbolu1991 commented:
@jukkar 
Regarding to PTP, I just added ptp clock shell commands support. You may help to review if you are interested in it:)
https://github.com/zephyrproject-rtos/zephyr/pull/89243

Thanks.

yangbolu1991 commented:
Updated to v2. Changes include,
- Added CONFIG_NET_GPTP_MONITOR_SYNC_STATUS to monitor real-time sync status, and print messages. Not only time offset and frequency offset now, there will be more users care about, like path delay, best clock changes and so on.

Thanks.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_mi.c:
```diff
@@ -822,28 +820,22 @@ static void gptp_update_local_port_clock(void)
-				 (unsigned long int)tm.nanosecond);
-		}
-
+#if defined(CONFIG_NET_GPTP_MONITOR_SYNC_STATUS)
+		NET_INFO("Set local clock %"PRIu64".%09u", tm.second, tm.nanosecond);
+#endif
```
I would prefer that we use the `IS_ENABLED()` macro so that the prints get compile tested properly. So better would be
```
if (IS_ENABLED(CONFIG_NET_GPTP_MONITOR_SYNC_STATUS)) {
	NET_INFO("Set local clock %"PRIu64".%09u", tm.second, tm.nanosecond);
}
```
Same comment to info print below.

yangbolu1991 replied:
A quick update on code for this. (I had thought with it disabled this code was useless to exist.)
Thanks.

clamattia replied:
FYI: In zephyr `IS_ENABLED` is preferred, whenever possible (all identifiers are present, it can compile). It will basically statically resolve to `true` or `false` and the compiler will eliminate the `dead` code.

yangbolu1991 replied:
Get it. Thanks a lot for the information:)
